src/handlers.h

```diff
@@ -14,5 +14,6 @@ void handle_code_cache(Deno* d, uint32_t cmd_id, const char* filename,
 void handle_timer_start(Deno* d, uint32_t cmd_id, uint32_t timer_id,
                         bool interval, uint32_t delay);
 void handle_timer_clear(Deno* d, uint32_t cmd_id, uint32_t timer_id);
+void handle_read_file_sync(Deno* d, uint32_t cmd_id, const char* filename);
```

Comment:
I guess it is appropriate to call the method `read_file_sync` since there currently is no way to use this function asynchronously.

But we should really try to avoid having two implementations  (sync and async) of every I/O function on the privileged side.

The JS layer can manage internally how it's going to present the result to the user; it can either block until recv() is called and return the value (sync), --or-- return nothing and resolve a promise when recv() gets called (async).

I could live with there being an `is_sync` field in the message, more as a hint to the backend, so that in the simple-sync model it can do I/O on the main thread and instead of offloading to a thread pool.

Reply:
I generally agree with having all the handlers be async and letting the JS layer choose... Two exceptions to that is
1. we're bootstrapping here - so I'm being quick and dirty with how these things are implemented - with the idea that we'll go back over  them again in the near future (like after #462 lands)
2. we've experienced in node where dispatching to thread pool for fs ops can be much slower than just synchronously doing those ops. I am worried that we may introduce the same here. Tho I suppose that was mostly caused by libuv's  use of eventfd?

Reply:
@ry 

1. On bootstrapping: I agree, I'm not saying you should can it. I was just trying to communicate the direction we should take after this lands. By looking at just the code iteself, people can't tell whether this is "just bootstrapping" or "the way to do things going forward". 
Maybe a add a comment?
This situation is not all that different to the `extern_version` patch the other day.

2. On thread pool performance: The libuv thread pool wasn't so great indeed, I can't say I don't worry about it at all. But my strong suspicion is that it's bogged down by the way it signals back from the worker to the main thread. Since the main thread blocks on epoll, there's three syscalls between pool work completing and the main thread picking it up (eventfd write, epoll, eventfd read). If you benchmark with a small read() op from a cache-hot file - in itself not a very expensive operation that takes one syscall - those three extra syscalls are going to be noticeable.

I think we'll fare much better if we use a lightweight primitive like a condition variable or a futex. It was also not that great that we didn't really have a good system to maintain ordering between operations when we need it, which in practice means that multiple thread pool threads can accept work items which cannot run in parallel, so they're now all serializing w.r.t one another.

Another source of perceived slowness is the relative inflexibility of the thread pool size. I think we can make it auto scale by running an extra low-priority thread which is normally sleeping (it never receives a CPU quantum because there are always higher-priority threads to run), but when it does wake up we take it as a hint that the other threads are now all blocked on I/O.

I also preemptively agree that my theories are unproven and we'll need to benchmark :)

Reply:
> I also preemptively agree that my theories are unproven and we'll need to benchmark :)

: )